### ISSUE TITLE
BEMDOM.detach doesn't remove _js_inited mod

### ISSUE THREAD

**Issue by h4** [bem/bem-core#1222]
I want to destruct block, but save dom node. @veged said that `BEMDOM.detach()` is correct way to do it in bem-core@2.7.0

But it doesn't work. 

Sample code https://github.com/h4/bem-detach-test/blob/master/desktop.blocks/test/test.js
Live demo http://h4.github.io/bem-detach-test/desktop.bundles/index/index.html

Click on 'Click me' button, block will be inited and background changed. After 1 second it call `BEMDOM.detach(this.domNode)` and I expect that background will be white. But it still light-green.

**Comment by veged** [comment]
cc @dfilatov @narqo

**Comment by korotovsky** [comment]
This issue is also affecs me

**Comment by veged** [comment]
@korotovsky could you please provide your real case for `BEMDOM.detach()` usage? cause in @h4 case we decide that there is no need in `BEMDOM.detach()` — I'm not gonna reject bug, but believe it has really rare effect

**Comment by korotovsky** [comment]
@veged I use i-bem with pjax library. So ... I have to call `BEMDOM.detach()` to detach all BEM blocks and keep DOM uninitialized before pjax will cache the current fragment and replace it with new one (received from server).

**Comment by veged** [comment]
ok, got it — see what we can do

**Comment by narqo** [comment]
@korotovsky could you describe your case in more details? What happens to this cached fragment (is it DOM fragment, right?) after it is replaced? 

It looks like recall of `BemDom.init()` won't work in most cases. Imagine block has a modifier which is possible only in runtime: for example you use `onSetMod` with this modifier and add some event handler within it. In case you try to reinit the block after the modifier is set, no event handlers will be added, because `onSetMod` won't be called.

**Comment by korotovsky** [comment]
@narqo What's your suggestion how to work with pjax (or SPA in general) and i-bem then?

**Comment by narqo** [comment]
@korotovsky I don't have any solutions now, but we need to look at your usecases in more details, so we can investigate the problem.

**Comment by dfilatov** [comment]
@korotovsky How is `detach` related to SPA? Why is `destruct` not enough?

**Comment by korotovsky** [comment]
@dfilatov Do you know [how pjax works](https://github.com/defunkt/jquery-pjax#overview)? It caches the current fragment (`#main-container` for example) to local private cache and when I press "back" button it restores the previous cached fragment and puts it to `#main-container`. That's why I can't use `destruct` because I have to keep DOM only uninitialized.

**Comment by dfilatov** [comment]
@korotovsky Does it cache string or DOM representation of the current fragment?

**Comment by korotovsky** [comment]
@dfilatov DOM.

**Comment by dfilatov** [comment]
@korotovsky So why do you need to `detach` in that case?

**Comment by korotovsky** [comment]
@dfilatov To uninitialize all event-listeners to prevent memory leaks. And another one is I have the chart library (morris.js). It loss some of its hover listeners after restoring from private cache, but I'm unable to reinitialize it because my BEM block (wrapper for morris.js chart) is already initialized.

**Comment by dfilatov** [comment]
@korotovsky It's weird you care of memory leaks for event listeners but don't care of memory leaks for DOM nodes.

**Comment by korotovsky** [comment]
> but doesn't care of memory leaks for DOM nodes.

@dfilatov Please, could you take a look into the pjax library? It has a `maxCacheLength` for its private cache: https://github.com/defunkt/jquery-pjax/blob/master/jquery.pjax.js#L804

And as I mentioned before, memory leaks is only one of my problems which I have when `BEMDOM.detach()` does not remove `*_js_inited` class.

UPDATE:

TL;DR: I'd like to cache only uninitialized DOM fragments. In my case this is the best way, because uninitialized fragment is the same as HTML fragment from server. The only thing that I need is - initialize BEM blocks after pjax will insert this fragment to page container.

**Comment by dfilatov** [comment]
@korotovsky Ok, I've got it.

**Comment by veged** [comment]
We can just fix removing of `*_js_inited` classes. However, problems described by @narqo still be here and you need to deal with them in code of concrete blocks.

**Comment by qfox** [comment]
Propose to add one more state: `_js_deinited` ;-)

**Comment by veged** [comment]
`_js_deinited` is the same as remove `_js_inited`

**Comment by dfilatov** [comment]
@h4 @isuvorov @korotovsky Do you use bem-components? If you do, bem-components doesn't support detaching with following reiniting.

**Comment by korotovsky** [comment]
@dfilatov No, we don't.

**Comment by veged** [comment]
since we use `BEMDOM.detach` only in some edge cases (for example in React wrapper) we do not need both: this consistency in modifiers and performance penalty in `className` modifications — so considering as wan't fix for now
